### ISSUE TITLE
Undefined variable false report in list/dict comprehension in formatstr

### ISSUE THREAD

**Issue by schneebuzz** [PyCQA/pylint#2491]
### Steps to reproduce
1. Use standard configuration
2. Lint the following code

```python
a = f'The ting: {[x for x in range(42)]}'
```

### Current behavior
`Undefined variable 'x' (undefined-variable`

### Expected behavior
No errors or warnings

### pylint --version output
```
pylint 2.1.1
astroid 2.0.4
Python 3.6.5 (v3.6.5:f59c0932b4, Mar 28 2018, 17:00:18) [MSC v.1900 64 bit (AMD64)]
```

Also see #1897 (partly a follow up)

**Comment by PCManticore** [comment]
I cannot reproduce this with the exact same versions:
```
pylint 2.1.1
astroid 2.0.4
Python 3.6.5 (default, May  7 2018, 10:07:07)
[GCC 4.2.1 Compatible Apple LLVM 8.1.0 (clang-802.0.42)]
```
```
************* Module a
a.py:1:0: C0111: Missing module docstring (missing-docstring)
a.py:1:0: C0103: Constant name "a" doesn't conform to UPPER_CASE naming style (invalid-name)

----------------------------------------------------------------------
Your code has been rated at -10.00/10 (previous run: -10.00/10, +0.00)
```

Can you reproduce it using your example? Or is there something missing from your posted example?

**Comment by PCManticore** [comment]
Closing for now, but let's reopen if it's still reproducing for you.

**Comment by PCManticore** [comment]
I managed to reproduce this by adding an additional line before the f-string.

**Comment by schneebuzz** [comment]
Yes, sorry that was not exactly the example to reproduce, I thought the lines I had above would not have any impact. Thanks for your effort on reproducing!

**Comment by sashgorokhov** [comment]
Have same bug, here my sample module code to reproduce:
```python
"""Foo bar"""

some_string = f"hello {[i for i in range(4)]}"
print(some_string)
```

```
pylint 2.2.2
astroid 2.1.0
Python 3.6.8 (tags/v3.6.8:3c6b436a57, Dec 24 2018, 00:16:47) [MSC v.1916 64 bit (AMD64)]
```

**Comment by WSLUser** [comment]
I get Undefined variable 'ODFParagraph' when I have odfpy installed with pip and imported OpenDocument (and odf.style, table, and text).

**Comment by kaddkaka** [comment]
Is this bug still present?

**Comment by DanielNoord** [comment]
@Pierre-Sassoulas I was checking for issues involving f-strings and found this. I can't reproduce the issues here. The examples only create a `unnecessary-comprehension` message but the `undefined-variable` message seems to be gone.
You might want to check for yourself, but I think this can be closed.

<img width="1037" alt="Schermafbeelding 2021-08-02 om 16 22 13" src="https://user-images.githubusercontent.com/13665637/127876658-59dcb7aa-d687-4670-864e-a98ee05d7ee5.png">

**Comment by Pierre-Sassoulas** [comment]
Thank you @DanielNoord !
